proof/refine/ARM/KHeap_R.thy

```diff
@@ -4288,4 +4288,52 @@ global_interpretation set_simple_ko: typ_at_pres "set_simple_ko C ptr ep"
 global_interpretation update_sk_obj_ref: typ_at_pres "update_sk_obj_ref C update ref new"
   unfolding typ_at_pres_def by wpsimp
 
+(* A standard (and active) scheduling context should have the following properties. They follow
+   from valid_sched_context' and sc_valid_refills as the following two lemmas show. *)
+abbreviation std_sc' where
+  "std_sc' ko \<equiv> 0 < scRefillCount ko \<and> scRefillMax ko \<le> length (scRefills ko) \<and> scRefillHead ko < scRefillMax ko"
+
+lemma std_sc'_normal:
+  "valid_sched_context' sc' s' \<Longrightarrow> 0 < scRefillMax sc' \<Longrightarrow> 0 < scRefillCount sc' \<Longrightarrow> std_sc' sc'"
+  by (clarsimp simp: valid_sched_context'_def)
+
+lemma sc_valid_refills_scRefillCount:
+  "\<lbrakk>sc_valid_refills sc; sc_relation sc n sc'\<rbrakk> \<Longrightarrow> 0 < scRefillCount sc'"
+  apply (clarsimp simp: valid_sched_context_def sc_relation_def)
+  apply (case_tac "scRefillCount sc'"; simp)
+  by (clarsimp simp: refills_map_def sc_valid_refills_def rr_valid_refills_def)
+
+lemma refill_hd_relation:
+  "sc_relation sc n sc' \<Longrightarrow> std_sc' sc' \<Longrightarrow>
+  refill_hd sc = refill_map (refillHd sc')"
+  apply (clarsimp simp: sc_relation_def refillHd_def refills_map_def valid_sched_context'_def hd_map)
+  apply (subst hd_map, clarsimp simp: wrap_slice_def)
+  apply (clarsimp simp: hd_wrap_slice)
+  done
+
+lemma sr_sc_refill_ready:
+  "\<lbrakk>sc_relation sc n sc'; std_sc' sc'\<rbrakk> \<Longrightarrow>
+  sc_refill_ready time sc = (rTime (refillHd sc') \<le> time + kernelWCETTicks)"
+  apply (subgoal_tac "rTime (refillHd sc') = r_time (refill_hd sc)")
+   apply (clarsimp simp: refill_ready_def state_relation_def ARM_H.kernelWCETTicks_def)
+  apply (clarsimp simp: refill_hd_relation)
+  apply (clarsimp simp: sc_relation_def refillHd_def refill_map_def)
+  done
+
+lemma sr_sc_refill_capacity:
+  "\<lbrakk>sc_relation sc n sc'; std_sc' sc'\<rbrakk> \<Longrightarrow>
+  refillsCapacity x (scRefills sc') (scRefillHead sc') = sc_refill_capacity x sc"
+  apply (clarsimp simp: refillsCapacity_def refill_capacity_def)
+  apply (clarsimp simp: refill_hd_relation)
+  apply (clarsimp simp: sc_relation_def refillHd_def refill_map_def)
+  done
+
+lemma sr_sc_refill_sufficient:
+  "\<lbrakk>sc_relation sc n sc'; std_sc' sc'\<rbrakk> \<Longrightarrow>
+  sufficientRefills x (scRefills sc') (scRefillHead sc') = sc_refill_sufficient x sc"
+  apply (frule (1) sr_sc_refill_capacity[where x=x])
+  apply (clarsimp simp: sufficientRefills_def refill_sufficient_def minBudget_def MIN_BUDGET_def
+                        ARM_H.kernelWCETTicks_def sc_relation_def)
+  done
+
```

Comment:
I think this block is an improvement over what existed previously, do you think it can be improved further?

Reply:
I think these lemmas belong to StateRelation rather than KHeap_R. And, I am basically repeating my comments earlier, but these are essentially duplication of what is already in StateRelation.thy, lines 260-287.

Apart from bundling some of the assumptions as abbreviation, `refill_hd_relation` is the same as `refills_heads_equal`. 

And, there is `refills_heads_equal_valid_sched_context’`, which more ore less gives you the identity of the head refill (both in `r_time` and `r_amount`) under `sc_relation` and `valid_sched_context’`. That effectively gives the equalities in readiness, sufficiency and capacity shown above. What is shown here probably suggests we can strengthen the `refills_heads_equal_valid_sched_context’` lemma.

So, I think we want to clean these up a bit better so that we won't leave both of them in different locations. Otherwise, one day we might end up working on a lemma that involves both, which can be pretty messy.

Reply:
Sorry to make you repeat yourself, thanks for the feedback, I'll make those changes.

Reply:
Thanks, Mitch. If you can look into consolidating these, that would be great.